CASMCMS-8841: Allow further granularity for controlling the boot progression for individual components. (#4618)
Cray-HPE/docs-csm: operations/boot_orchestration/Tools_for_Resolving_Boot_Issues.md

A number of tools can be used to analyze and debug issues encountered during the compute node boot process. The underlying issue and symptoms dictate the type of tool required.

## BOS

BOS component records offer a history of operations that have been applied to a node.

```bash
cray bos components describe x3000c0s19b4n0
```

> Additionally, nodes that are unable to boot may also be disabled. BOS will not apply power actions to disabled nodes,

Reviewer: Is there any other BOS operation that can clear the "enabled = False" state, or does the administrator have sole control over that?

Author: At the end of the day, the component record is served up by an API, so any software or user could change this, however, in practice, the only things that toggle the enabled/disabled flag on the BOS level are:
- Nodes completing boot will toggle from enabled to disabled
- New BOS boot sessions that include the affected components will move from disabled to enabled
- Any other user who decides to enable or disable operation on a given BOS component.

Reviewer 2: @don-bahls-hpe , note that in this case enabled = True means that BOS is going to operate on it. enabled = False means BOS is not going to operate on it. But this is a dynamic state. Nodes are enabled when a new BOS session is created. That is the first step.

> allowing an administrator to debug the node environment interactively.

Reviewer 2: Do we need to say anything about the node being disable, so long as it does not become a part of a new session?

I suppose an alternate way to disable nodes is to do so in HSM. That would also keep the node from being operated on. I think it's more dangerous because HSM disabling has to be manually undone whereas just making the node a part of a new BOS session re-enables it. From the debugger's point of view though, the reverse is true. HSM-disablement is safer because it cannot be accidentally undone.

Author: I had attempted to view the problem as what we could accomplish in terms of cfs specifically, but this is probably the fine grained hammer that we need. Besides; not to hash up previous regretful skeletons, but HSM's disable flag in the PCS world is on the honor system 😆

Reviewer 2: Doesn't BOS always honor it though, or did we change that when we converted to PCS?

I missed the connection between this documentation and it being specifically related to CFS. Could you explain?

```bash
cray bos components update --enabled False x3000c0s19b4n0
```

> Once the problem has been isolated and fixed, re-enable the node in BOS. Note: when re-enabling a node, BOS may first power cycle it to return it to working order, depending on elapsed time and BOS options.

Reviewer 2: This may make sessions look weird. For example, if I asked to shut down all of the nodes, but one didn't. I'd disable it, so I could go poke at it. Once the other nodes shut down, the session completes. Re-enabling this guy will cause the session to start running again, assuming it hasn't been cleaned up. We knew that would be a problem in the way we designed it. So these are probably idle musings.

A number of tools can be used to analyze and debug issues encountered during the compute node boot process. The underlying issue and symptoms dictate the type of tool required.

> ## `BOS`

Reviewer 3: ```suggestion
## BOS
```

> Boot orchestration services offers records of actions that have been applied to a node through their component records.

Reviewer 3: ```suggestion
BOS component records offer a history of operations that have been applied to a node.
```

A history of applied actions and future actions can be used to determine corrective actions and controlling actions that

> enables administrators to introspect issues on a per node level.

Reviewer 3: ```suggestion
```

```bash
cray bos components describe x3000c0s19b4n0
```

> Additionally, nodes that are unable to boot may also be disabled to prevent further BOS power actions from being applied
> to a node. This is helpful if BOS is currently retrying power operations on a specific node, and you need to introspect
> the node environment interactively.

Reviewer 3: ```suggestion
Additionally, nodes that are unable to boot may also be disabled. BOS will not apply power actions to disabled nodes,
allowing an administrator to debug the node environment interactively.
```

```bash
cray bos components update --enabled False x3000c0s19b4n0
```

> Once the problem has been isolated and fixed, you can return the node back over to BOS so that it may resume its boot
> session. Note: when re-enabling a node, BOS may first power cycle it to return it to working order depending on elapsed
> time and BOS options.

Reviewer 3: ```suggestion
Once the problem has been isolated and fixed, re-enable the node in BOS. Note: when re-enabling a node, BOS may first power cycle it to return it to working order, depending on elapsed time and BOS options.
```
